v2/permissions/cache.go

```diff
@@ -0,0 +1,125 @@
+package permissions
+
+import (
+	"context"
+	"github.com/ONSdigital/log.go/v2/log"
+	"sync"
+	"time"
+)
+
+// Compiler check to ensure CachingStore implements the Store interface.
+var _ Store = (*CachingStore)(nil)
+
+// CachingStore is a permissions store implementation that caches permission data in memory.
+type CachingStore struct {
+	underlyingStore     Store
+	cachedBundle        *Bundle
+	closing             chan struct{}
+	expiryCheckerClosed chan struct{}
+	cacheUpdaterClosed  chan struct{}
+	lastUpdated         time.Time
+	mutex               sync.Mutex
+}
+
+// NewCachingStore contructs a new instance of CachingStore
+func NewCachingStore(underlyingStore Store) *CachingStore {
+	return &CachingStore{
+		underlyingStore:     underlyingStore,
+		closing:             make(chan struct{}),
+		expiryCheckerClosed: make(chan struct{}),
+		cacheUpdaterClosed:  make(chan struct{}),
+	}
+}
+
+// GetPermissionsBundle returns the cached permission data, or an error if it's not cached.
+func (c *CachingStore) GetPermissionsBundle(ctx context.Context) (*Bundle, error) {
```

Comment:
Does this need to return an error if the bundle is not cached? Could it return nil instead? I assume that if its not cached the next step is to retrieve the data and cache it? If so err feels wrong as its technically not an error - its a valid state.

Reply:
Yes this is by design. The cache should always be populated. If it's not, it means that the (background) cache update has failed, and the existing cache data has expired.